aiida/tools/graph/graph_traversers.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Module for functions to traverse AiiDA graphs."""
+
+from __future__ import absolute_import
+from numpy import inf
+
+
+def traverse_graph(starting_pks, max_iterations=None, get_links=False, **kwargs):
+    """
+    This function will return the set of all nodes that can be connected
+    to a list of initial nodes through any sequence of specified authorized
+    links and directions.
+
+    :type starting_pks:  list
+    :param starting_pks:
+        A list with the (valid) pks of all starting nodes.
+
+    :type max_iterations: int or None
+    :param max_iterations:
+        The number of iterations to apply the set of rules (a value of 'None' will
+        iterate until no new nodes are added).
+
+    :type get_links:  bool
+    :param get_links:
+        Pass True to also return the links between all nodes (found + initial).
+
+    :type traverse_rules: aiida.common.links.GraphTraversalRule
+    :param traverse_rules:
```

Comment:
This looks a bit confusing. I guess there is no parameter `traversal_rules`, but there can be a number of traversal rules pass as kwargs, correct?
Perhaps rather describe this in words (not sure whether there is a standardized syntax for this).

Reply:
Yes, this is right, there is no single traverse_rules but the code will look into the provided `kwargs` for all the specific rules (`input_calc_forward`, `create_backward`, etc), which **must** be present (each of them). Should I explicitly put all of them maybe? It feels like it defeats a bit the purpose of having them be defined separately in `aiida.common.links`, or at least that it is adding some undesired repetition, but maybe this points towards a more fundamental problem with this approach that should be dealt with in another occasion.

Reply:
I see, I didn't realize that all of them need to be present.
If that is the case, it is probably indeed easier to simply pass a `traversal_rules` dictionary that contains these lists.
I guess that would not be a big change?

Reply:
Mmm, at first glance I would say that could be better. I will do the change with the next set of reviews so as not to disturb the PR in case somebody started reviewing.

Reply:
Re-structured a bit the graph traversal interface. See upcoming comment after I push the new changes.